lib/Dialect/FIRRTL/FIRRTLOps.cpp

```diff
@@ -842,8 +883,97 @@ void InstanceOp::build(OpBuilder &builder, OperationState &result,
       removeElementsAtIndices<Type>(resultTypes, resultsToErase);
 
-  build(builder, result, newResultTypes, existingInstance.moduleNameAttr(),
-        existingInstance.nameAttr(), existingInstance.annotationsAttr());
+  build(builder, result, newResultTypes, existingInstance->getOperands(),
+        existingInstance->getAttrs());
+}
+
+static ParseResult parseInstanceOp(OpAsmParser &parser,
+                                   OperationState &result) {
+  FlatSymbolRefAttr moduleNameAttr;
+  if (parser.parseAttribute(moduleNameAttr,
+                            parser.getBuilder().getType<NoneType>(),
+                            "moduleName", result.attributes))
+    return failure();
+
+  if (parseElideAnnotations(parser, result.attributes))
+    return failure();
+
+  SmallVector<Type, 4> resultsTypes;
+  SmallVector<Attribute, 4> resultsAttrs;
+  if (succeeded(parser.parseOptionalColon())) {
+    do {
+      resultsTypes.emplace_back();
+      NamedAttrList annotation;
+      if (parser.parseType(resultsTypes.back()) ||
+          parser.parseOptionalAttrDict(annotation))
+        return failure();
+
+      resultsAttrs.push_back(
+          DictionaryAttr::get(result.getContext(), annotation));
+    } while (succeeded(parser.parseOptionalComma()));
+  }
+
+  result.addTypes(resultsTypes);
+  result.addAttribute("portAnnotations",
+                      ArrayAttr::get(result.getContext(), resultsAttrs));
+  return success();
+}
+
+static void print(OpAsmPrinter &p, InstanceOp op) {
+  p << "firrtl.instance";
+  p << ' ';
+  p.printAttributeWithoutType(op.moduleNameAttr());
+  p << ' ';
+  printElideAnnotations(p, op, op.getOperation()->getAttrDictionary(),
+                        {"moduleName", "portAnnotations"});
+
+  if (op.getNumResults()) {
+    p << " : ";
+    auto portTypes = op.getResultTypes();
+    auto portAnnotations = op->getAttrOfType<ArrayAttr>("portAnnotations");
+
+    for (unsigned i = 0, e = op.getNumResults(); i < e; ++i) {
+      if (i > 0)
+        p << ", ";
+
+      p.printType(portTypes[i]);
+      if (portAnnotations)
+        p.printOptionalAttrDict(
+            portAnnotations[i].cast<DictionaryAttr>().getValue());
+    }
+  }
+}
+
+DictionaryAttr InstanceOp::getPortAnnotation(unsigned portIdx) {
+  if (portIdx >= getNumResults())
+    return emitOpError("invalid result index"), nullptr;
+
+  if (auto attr = (*this)->getAttrOfType<ArrayAttr>("portAnnotations"))
+    return attr[portIdx].cast<DictionaryAttr>();
+  return nullptr;
+}
+
+bool InstanceOp::setPortAnnotation(unsigned portIdx,
+                                   DictionaryAttr annotation) {
+  if (portIdx >= getNumResults())
+    return emitOpError("invalid result index"), false;
+
+  mlir::function_like_impl::detail::setArgResAttrDict(
+      (*this), "portAnnotations", getNumResults(), portIdx, annotation);
+  return true;
+}
+
+bool InstanceOp::setPortAnnotation(StringRef portName,
+                                   DictionaryAttr annotation) {
```

Comment:
Please don't add this helper method.  The "getReferencedModule()" call is O(n) in the size of the CircuitOp.  We should do everything in terms of port indices, to avoid algorithmic cliffs.